tests/bench_timers: improve test to work on some SAM based boards

fedepell opened the pull request:
### Contribution description
This PR does some small improvements to the bench_timers test to make it work on some SAM based boards and probably also improve test portability to other boards.
Specifically:
-) specific SAM boards definition with only one timer are added to Makefile so correct timer is used
-) before calling `timer_init` on a timer this is first stopped with `timer_stop` otherwise for most of SAM based CPUs the init will not have any effect since if the timer has already been inited (and it is at boot) then initialization would be skipped and test would block undefinitely waiting for the callback to be called. 
Ie see code:
```
/**
 * @brief Setup the given timer
 */
int timer_init(tim_t dev, unsigned long freq, timer_cb_t cb, void *arg)
{
   ...
        if (TIMER_0_DEV.CTRLA.bit.ENABLE) {
            return 0;
        }
```
-) Default minimum time for relative timer test (`TEST_MIN_REL`) is set to `TEST_MIN` (which is the minimum for absolute alarms) as the previous 0 results in timer shooting off before the callback is even set and test blocking. It sounds reasonable that the minimum for the relative test should be the same as the one for absolute by the user. It can anyhow, as before, be overridden explicitly if needed.

### Testing procedure
In tests/bench_timers:
`BOARD=XYZ make all flash`
And then see the test output (after a while) on the serial

### Issues/PRs references
Addresses problems of test execution explained in #10523 (even if doesn't address the timer driver problem).

benpicco commented:
> -) before calling timer_init on a timer this is first stopped with timer_stop otherwise for most of SAM based CPUs the init will not have any effect since if the timer has already been inited (and it is at boot) then initialization would be skipped and test would block undefinitely waiting for the callback to be called.

Instead of working around that, why not just remove the check in `timer_init()` and replace it with a `timer_stop()` ?

fedepell commented:
@benpicco thanks for checking the code! I first wanted to do that too indeed as was sounding more to solve the problem than a workaround. 
But then I opted for changing the test since changing this in `timer_init()` would change the behaviour of that function quite a bit as now the timer will never get overwritten until it is running (and what I found also not so polite is that it will even return 0, therefore not notify anything!), while with the change it would always use the last one. 
While the second one may be more correct (or at least getting an error to tell me that it was not inited) it was sounding as a big behaviour change that may need more discussion and a separate PR in case (or your PR of SAM timers rewrite :) ) and this PR would address simply the benchmark test execution by itself.

I checked a few others CPUs now and in general most of them indeed either overwrite the previous timer or return an error if it is re-inited. Maybe this is something that could be uniformed among all the CPUs indeed.

fedepell commented:
@gebart : mind giving a check and maybe a push? thanks! :)

fedepell commented:
@dylad : could you in case give it a try with your boards as you have some SAMs? (I tried on SAML21 and had feedback for SAMD21 and SAMR21) Thanks!

dylad commented:
I don't know much about timers stuff but here are the results of bench_timers for SAML10:

```
2019-05-08 18:45:55,980 - INFO # main(): This is RIOT! (Version: 2019.04-devel-838-g9b87f-bench_timers_saml)
2019-05-08 18:45:55,981 - INFO # 
2019-05-08 18:45:55,983 - INFO # Statistical benchmark for timers
2019-05-08 18:45:55,991 - INFO # Running timer test with seed 123 using Tiny Mersenne Twister PRNG.
2019-05-08 18:45:55,992 - INFO # TEST_MIN = 16
2019-05-08 18:45:55,994 - INFO # TEST_MAX = 128
2019-05-08 18:45:55,995 - INFO # TEST_MIN_REL = 16
2019-05-08 18:45:55,997 - INFO # TEST_MAX_REL = 128
2019-05-08 18:45:55,998 - INFO # TEST_NUM = 113
2019-05-08 18:45:56,001 - INFO # log2(TEST_NUM - 1) = 6
2019-05-08 18:45:56,004 - INFO # state vector elements per variant = 7
2019-05-08 18:45:56,006 - INFO # number of variants = 8
2019-05-08 18:45:56,009 - INFO # sizeof(state) = 32 bytes
2019-05-08 18:45:56,012 - INFO # state vector total memory usage = 1792 bytes
2019-05-08 18:45:56,018 - INFO # TIM_TEST_DEV = 0, TIM_TEST_FREQ = 1000000, TIM_TEST_CHAN = 0
2019-05-08 18:45:56,022 - INFO # TIM_REF_DEV  = 0, TIM_REF_FREQ  = 1000000
2019-05-08 18:45:56,023 - INFO # USE_REFERENCE = 1
2019-05-08 18:45:56,030 - INFO # TEST_PRINT_INTERVAL_TICKS = 30000000
2019-05-08 18:45:56,032 - INFO # Expected error variance due to truncation in tick conversion: 0
2019-05-08 18:45:56,035 - INFO # Calibrating spin delay...
2019-05-08 18:45:56,036 - INFO # spin_max = 38
2019-05-08 18:45:56,039 - INFO # Estimating benchmark overhead...
2019-05-08 18:45:56,284 - INFO # overhead_target = 18 (s2 = 0)
2019-05-08 18:45:56,287 - INFO # overhead_read = 18 (s2 = 0)
2019-05-08 18:45:56,288 - INFO # exp_mask = 00000007
2019-05-08 18:45:56,290 - INFO # max interval = 255
2019-05-08 18:45:56,505 - INFO # 
2019-05-08 18:45:56,508 - INFO # Context before hardfault:
2019-05-08 18:45:56,509 - INFO #    r0: 0x00000000
2019-05-08 18:45:56,511 - INFO #    r1: 0x00000004
2019-05-08 18:45:56,513 - INFO #    r2: 0xe000ed00
2019-05-08 18:45:56,514 - INFO #    r3: 0x00000010
2019-05-08 18:45:56,516 - INFO #   r12: 0x00000000
2019-05-08 18:45:56,518 - INFO #    lr: 0x00000a2f
2019-05-08 18:45:56,520 - INFO #    pc: 0x0000061c
2019-05-08 18:45:56,521 - INFO #   psr: 0x01000000
2019-05-08 18:45:56,522 - INFO # 
2019-05-08 18:45:56,522 - INFO # Misc
2019-05-08 18:45:56,524 - INFO # EXC_RET: 0xfffffffd
2019-05-08 18:45:56,528 - INFO # Attempting to reconstruct state for debugging...
2019-05-08 18:45:56,529 - INFO # In GDB:
2019-05-08 18:45:56,531 - INFO #   set $pc=0x61c
2019-05-08 18:45:56,532 - INFO #   frame 0
2019-05-08 18:45:56,532 - INFO #   bt
```
SAML21:
```
2019-05-08 18:48:55,577 - INFO # main(): This is RIOT! (Version: 2019.04-devel-838-g9b87f-bench_timers_saml)
2019-05-08 18:48:55,577 - INFO # 
2019-05-08 18:48:55,580 - INFO # Statistical benchmark for timers
2019-05-08 18:48:55,587 - INFO # Running timer test with seed 123 using Tiny Mersenne Twister PRNG.
2019-05-08 18:48:55,588 - INFO # TEST_MIN = 16
2019-05-08 18:48:55,591 - INFO # TEST_MAX = 128
2019-05-08 18:48:55,593 - INFO # TEST_MIN_REL = 16
2019-05-08 18:48:55,594 - INFO # TEST_MAX_REL = 128
2019-05-08 18:48:55,594 - INFO # TEST_NUM = 113
2019-05-08 18:48:55,596 - INFO # log2(TEST_NUM - 1) = 6
2019-05-08 18:48:55,600 - INFO # state vector elements per variant = 7
2019-05-08 18:48:55,602 - INFO # number of variants = 8
2019-05-08 18:48:55,604 - INFO # sizeof(state) = 32 bytes
2019-05-08 18:48:55,608 - INFO # state vector total memory usage = 1792 bytes
2019-05-08 18:48:55,614 - INFO # TIM_TEST_DEV = 0, TIM_TEST_FREQ = 1000000, TIM_TEST_CHAN = 0
2019-05-08 18:48:55,618 - INFO # TIM_REF_DEV  = 0, TIM_REF_FREQ  = 1000000
2019-05-08 18:48:55,619 - INFO # USE_REFERENCE = 1
2019-05-08 18:48:55,623 - INFO # TEST_PRINT_INTERVAL_TICKS = 30000000
2019-05-08 18:48:55,628 - INFO # Expected error variance due to truncation in tick conversion: 0
2019-05-08 18:48:55,631 - INFO # Calibrating spin delay...
2019-05-08 18:48:55,632 - INFO # spin_max = 47
2019-05-08 18:48:55,635 - INFO # Estimating benchmark overhead...
2019-05-08 18:48:55,875 - INFO # overhead_target = 14 (s2 = 0)
2019-05-08 18:48:55,878 - INFO # overhead_read = 14 (s2 = 0)
2019-05-08 18:48:55,880 - INFO # exp_mask = 00000007
2019-05-08 18:48:55,881 - INFO # max interval = 255
2019-05-08 18:50:01,758 - INFO # Elapsed time:
2019-05-08 18:50:01,761 - INFO #         Reference: 4294
2019-05-08 18:50:01,763 - INFO #  Timer under test: 4294
2019-05-08 18:50:01,765 - INFO #  Wall clock (RTT): 0
2019-05-08 18:50:01,768 - INFO # ------------- BEGIN STATISTICS --------------
2019-05-08 18:50:01,772 - INFO # ===== Reference timer statistics =====
2019-05-08 18:50:01,776 - INFO # Limits: mean: [-10, 10], variance: [0, 16]
2019-05-08 18:50:01,783 - INFO # Target error (actual trigger time - expected trigger time), in reference timer ticks
2019-05-08 18:50:01,790 - INFO # positive: timer under test is late, negative: timer under test is early
2019-05-08 18:50:01,792 - INFO # === timer_set running ===
2019-05-08 18:50:01,799 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:01,805 - INFO #   16 -   17:     5613     23182         2845      3     5     4      0
2019-05-08 18:50:01,812 - INFO #   18 -   19:     5607     11960         2839      1     3     2      0
2019-05-08 18:50:01,819 - INFO #   20 -   23:     5644      2121         4439     -1     4     0      0
2019-05-08 18:50:01,825 - INFO #   24 -   31:     5570     12678         4591      0     4     2      0
2019-05-08 18:50:01,832 - INFO #   32 -   47:     5590     13630         2450      2     3     2      0
2019-05-08 18:50:01,839 - INFO #   48 -   79:     5642     13797         2513      2     3     2      0
2019-05-08 18:50:01,845 - INFO #   80 -  143:     4304     10473         1865      2     3     2      0
2019-05-08 18:50:01,852 - INFO #       TOTAL     37970     87841        63664     -1     5     2      1
2019-05-08 18:50:01,855 - INFO # === timer_set resched ===
2019-05-08 18:50:01,861 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:01,868 - INFO #   16 -   17:     5435     22417         2757      3     5     4      0
2019-05-08 18:50:01,875 - INFO #   18 -   19:     5483     11673         2731      1     3     2      0
2019-05-08 18:50:01,881 - INFO #   20 -   23:     5646      2103         4357     -1     4     0      0
2019-05-08 18:50:01,888 - INFO #   24 -   31:     5660     12684         4499      0     4     2      0
2019-05-08 18:50:01,895 - INFO #   32 -   47:     5586     13661         2487      2     3     2      0
2019-05-08 18:50:01,901 - INFO #   48 -   79:     5607     13681         2467      2     3     2      0
2019-05-08 18:50:01,908 - INFO #   80 -  143:     4233     10310         1842      2     3     2      0
2019-05-08 18:50:01,915 - INFO #       TOTAL     37650     86529        61768     -1     5     2      1
2019-05-08 18:50:01,918 - INFO # === timer_set stopped ===
2019-05-08 18:50:01,924 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:01,934 - INFO #   16 -   17:     5651    -36287         4121     -7    -3    -6      0
2019-05-08 18:50:01,938 - INFO #   18 -   19:     5657    -32424         6162     -6    -3    -5      1
2019-05-08 18:50:01,944 - INFO #   20 -   23:     5454    -20878         4817     -5    -3    -3      0
2019-05-08 18:50:01,951 - INFO #   24 -   31:     5481    -21924            0     -4    -4    -4      0
2019-05-08 18:50:01,958 - INFO #   32 -   47:     5652    -22608            0     -4    -4    -4      0
2019-05-08 18:50:01,964 - INFO #   48 -   79:     5434    -21736            0     -4    -4    -4      0
2019-05-08 18:50:01,971 - INFO #   80 -  143:     4289    -17156            0     -4    -4    -4      0
2019-05-08 18:50:01,978 - INFO #       TOTAL     37618   -173013       121361     -7    -3    -4      3
2019-05-08 18:50:01,981 - INFO # === timer_set stopped, resched ===
2019-05-08 18:50:01,988 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:01,994 - INFO #   16 -   17:     5572    -35815         4014     -7    -3    -6      0
2019-05-08 18:50:02,001 - INFO #   18 -   19:     5598    -32125         6174     -6    -3    -5      1
2019-05-08 18:50:02,008 - INFO #   20 -   23:     5564    -21280         4882     -5    -3    -3      0
2019-05-08 18:50:02,014 - INFO #   24 -   31:     5544    -22176            0     -4    -4    -4      0
2019-05-08 18:50:02,021 - INFO #   32 -   47:     5525    -22100            0     -4    -4    -4      0
2019-05-08 18:50:02,028 - INFO #   48 -   79:     5562    -22248            0     -4    -4    -4      0
2019-05-08 18:50:02,034 - INFO #   80 -  143:     4278    -17112            0     -4    -4    -4      0
2019-05-08 18:50:02,042 - INFO #       TOTAL     37643   -172856       119482     -7    -3    -4      3
2019-05-08 18:50:02,045 - INFO # === timer_set_absolute running ===
2019-05-08 18:50:02,051 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:02,058 - INFO #   16 -   17:     5528      1055         2779     -1     1     0      0
2019-05-08 18:50:02,065 - INFO #   18 -   19:     5555    -10075         6260     -3    -1    -1      1
2019-05-08 18:50:02,071 - INFO #   20 -   23:     5618    -19026         4916     -5     0    -3      0
2019-05-08 18:50:02,078 - INFO #   24 -   31:     5600     -8114         4800     -4     0    -1      0
2019-05-08 18:50:02,085 - INFO #   32 -   47:     5492     -7200         1708     -2    -1    -1      0
2019-05-08 18:50:02,091 - INFO #   48 -   79:     5500     -7202         1700     -2    -1    -1      0
2019-05-08 18:50:02,098 - INFO #   80 -  143:     4191     -5524         1333     -2    -1    -1      0
2019-05-08 18:50:02,105 - INFO #       TOTAL     37484    -56086        62603     -5     1    -1      1
2019-05-08 18:50:02,108 - INFO # === timer_set_absolute resched ===
2019-05-08 18:50:02,115 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:02,121 - INFO #   16 -   17:     5707      1024         2817     -1     1     0      0
2019-05-08 18:50:02,128 - INFO #   18 -   19:     5629    -10165         6246     -3    -1    -1      1
2019-05-08 18:50:02,135 - INFO #   20 -   23:     5582    -18948         4918     -5     0    -3      0
2019-05-08 18:50:02,141 - INFO #   24 -   31:     5588     -8072         4680     -4     0    -1      0
2019-05-08 18:50:02,148 - INFO #   32 -   47:     5619     -7384         1764     -2    -1    -1      0
2019-05-08 18:50:02,155 - INFO #   48 -   79:     5655     -7471         1816     -2    -1    -1      0
2019-05-08 18:50:02,162 - INFO #   80 -  143:     4272     -5564         1292     -2    -1    -1      0
2019-05-08 18:50:02,168 - INFO #       TOTAL     38052    -56580        62453     -5     1    -1      1
2019-05-08 18:50:02,172 - INFO # === timer_set_absolute stopped ===
2019-05-08 18:50:02,178 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:02,185 - INFO #   16 -   17:     5423    -34880         3818     -7    -3    -6      0
2019-05-08 18:50:02,192 - INFO #   18 -   19:     5549    -31920         6074     -6    -3    -5      1
2019-05-08 18:50:02,198 - INFO #   20 -   23:     5581    -21318         4869     -5    -3    -3      0
2019-05-08 18:50:02,205 - INFO #   24 -   31:     5616    -22464            0     -4    -4    -4      0
2019-05-08 18:50:02,212 - INFO #   32 -   47:     5570    -22280            0     -4    -4    -4      0
2019-05-08 18:50:02,218 - INFO #   48 -   79:     5453    -21812            0     -4    -4    -4      0
2019-05-08 18:50:02,225 - INFO #   80 -  143:     4296    -17184            0     -4    -4    -4      0
2019-05-08 18:50:02,232 - INFO #       TOTAL     37488   -171858       117151     -7    -3    -4      3
2019-05-08 18:50:02,236 - INFO # === timer_set_absolute stopped, resched ===
2019-05-08 18:50:02,243 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:02,250 - INFO #   16 -   17:     5586    -35762         4058     -7    -3    -6      0
2019-05-08 18:50:02,256 - INFO #   18 -   19:     5595    -32156         6066     -6    -3    -5      1
2019-05-08 18:50:02,263 - INFO #   20 -   23:     5604    -21493         4963     -5    -3    -3      0
2019-05-08 18:50:02,269 - INFO #   24 -   31:     5654    -22616            0     -4    -4    -4      0
2019-05-08 18:50:02,276 - INFO #   32 -   47:     5639    -22556            0     -4    -4    -4      0
2019-05-08 18:50:02,283 - INFO #   48 -   79:     5644    -22576            0     -4    -4    -4      0
2019-05-08 18:50:02,289 - INFO #   80 -  143:     4303    -17212            0     -4    -4    -4      0
2019-05-08 18:50:02,297 - INFO #       TOTAL     38025   -174371       119306     -7    -3    -4      3
2019-05-08 18:50:02,300 - INFO # ===== introspective statistics =====
2019-05-08 18:50:02,304 - INFO # Limits: mean: [-10, 10], variance: [0, 16]
2019-05-08 18:50:02,312 - INFO # self-referencing error (TUT time elapsed - expected TUT interval), in timer under test ticks
2019-05-08 18:50:02,321 - INFO # positive: timer target handling is slow, negative: TUT is dropping ticks or triggering callback early
2019-05-08 18:50:02,328 - INFO # function              count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:50:02,328 - INFO # 
2019-05-08 18:50:02,335 - INFO #          timer_set   150881   -801916       864951     -9     0    -5      5
2019-05-08 18:50:02,343 - INFO #            running    37970   -118974        61926     -6     0    -3      1
2019-05-08 18:50:02,350 - INFO #            resched    37650   -118651        61284     -6     0    -3      1
2019-05-08 18:50:02,357 - INFO #            stopped    37618   -282150        44328     -9    -5    -7      1
2019-05-08 18:50:02,364 - INFO #   stopped, resched    37643   -282141        44081     -9    -5    -7      1
2019-05-08 18:50:02,365 - INFO # 
2019-05-08 18:50:02,372 - INFO # timer_set_absolute   151049  -1089526        43984     -9    -4    -7      0
2019-05-08 18:50:02,379 - INFO #            running    37484   -260138        84987     -9    -4    -6      2
2019-05-08 18:50:02,386 - INFO #            resched    38052   -263683        84687     -9    -4    -6      2
2019-05-08 18:50:02,394 - INFO #            stopped    37488   -280808        43360     -9    -5    -7      1
2019-05-08 18:50:02,401 - INFO #   stopped, resched    38025   -284897        43984     -9    -5    -7      1
2019-05-08 18:50:02,405 - INFO # -------------- END STATISTICS ---------------
2019-05-08 18:51:08,154 - INFO # Elapsed time:
2019-05-08 18:51:08,156 - INFO #         Reference: 4294
2019-05-08 18:51:08,158 - INFO #  Timer under test: 4294
2019-05-08 18:51:08,160 - INFO #  Wall clock (RTT): 0
2019-05-08 18:51:08,164 - INFO # ------------- BEGIN STATISTICS --------------
2019-05-08 18:51:08,167 - INFO # ===== Reference timer statistics =====
2019-05-08 18:51:08,171 - INFO # Limits: mean: [-10, 10], variance: [0, 16]
2019-05-08 18:51:08,179 - INFO # Target error (actual trigger time - expected trigger time), in reference timer ticks
2019-05-08 18:51:08,185 - INFO # positive: timer under test is late, negative: timer under test is early
2019-05-08 18:51:08,188 - INFO # === timer_set running ===
2019-05-08 18:51:08,194 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,201 - INFO #   16 -   17:    11177     46158         5623      3     5     4      0
2019-05-08 18:51:08,208 - INFO #   18 -   19:    11071     23520         5563      1     3     2      0
2019-05-08 18:51:08,214 - INFO #   20 -   23:    11276      4205         8877     -1     4     0      0
2019-05-08 18:51:08,221 - INFO #   24 -   31:    11168     25345         9018      0     4     2      0
2019-05-08 18:51:08,228 - INFO #   32 -   47:    11254     27443         4935      2     3     2      0
2019-05-08 18:51:08,234 - INFO #   48 -   79:    11189     27310         4932      2     3     2      0
2019-05-08 18:51:08,241 - INFO #   80 -  143:     8486     20620         3648      2     3     2      0
2019-05-08 18:51:08,248 - INFO #       TOTAL     75621    174601       126502     -1     5     2      1
2019-05-08 18:51:08,250 - INFO # === timer_set resched ===
2019-05-08 18:51:08,257 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,264 - INFO #   16 -   17:    10981     45332         5546      3     5     4      0
2019-05-08 18:51:08,270 - INFO #   18 -   19:    10999     23326         5520      1     3     2      0
2019-05-08 18:51:08,277 - INFO #   20 -   23:    11327      4271         9101     -1     4     0      0
2019-05-08 18:51:08,284 - INFO #   24 -   31:    11291     25462         8985      0     4     2      0
2019-05-08 18:51:08,290 - INFO #   32 -   47:    11223     27341         4893      2     3     2      0
2019-05-08 18:51:08,297 - INFO #   48 -   79:    11116     27122         4890      2     3     2      0
2019-05-08 18:51:08,304 - INFO #   80 -  143:     8490     20678         3696      2     3     2      0
2019-05-08 18:51:08,311 - INFO #       TOTAL     75427    173532       124753     -1     5     2      1
2019-05-08 18:51:08,313 - INFO # === timer_set stopped ===
2019-05-08 18:51:08,320 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,326 - INFO #   16 -   17:    11164    -71699         8195     -7    -3    -6      0
2019-05-08 18:51:08,333 - INFO #   18 -   19:    11231    -64438        12242     -6    -3    -5      1
2019-05-08 18:51:08,340 - INFO #   20 -   23:    10982    -42048         9741     -5    -3    -3      0
2019-05-08 18:51:08,347 - INFO #   24 -   31:    11034    -44136            0     -4    -4    -4      0
2019-05-08 18:51:08,353 - INFO #   32 -   47:    11195    -44780            0     -4    -4    -4      0
2019-05-08 18:51:08,360 - INFO #   48 -   79:    10969    -43876            0     -4    -4    -4      0
2019-05-08 18:51:08,367 - INFO #   80 -  143:     8515    -34060            0     -4    -4    -4      0
2019-05-08 18:51:08,374 - INFO #       TOTAL     75090   -345037       239967     -7    -3    -4      3
2019-05-08 18:51:08,377 - INFO # === timer_set stopped, resched ===
2019-05-08 18:51:08,383 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,390 - INFO #   16 -   17:    11157    -71763         7904     -7    -3    -6      0
2019-05-08 18:51:08,397 - INFO #   18 -   19:    10995    -63204        12016     -6    -3    -5      1
2019-05-08 18:51:08,403 - INFO #   20 -   23:    11214    -42882         9878     -5    -3    -3      0
2019-05-08 18:51:08,410 - INFO #   24 -   31:    11218    -44872            0     -4    -4    -4      0
2019-05-08 18:51:08,417 - INFO #   32 -   47:    11071    -44284            0     -4    -4    -4      0
2019-05-08 18:51:08,423 - INFO #   48 -   79:    11014    -44056            0     -4    -4    -4      0
2019-05-08 18:51:08,430 - INFO #   80 -  143:     8492    -33968            0     -4    -4    -4      0
2019-05-08 18:51:08,437 - INFO #       TOTAL     75161   -345029       236930     -7    -3    -4      3
2019-05-08 18:51:08,440 - INFO # === timer_set_absolute running ===
2019-05-08 18:51:08,447 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,454 - INFO #   16 -   17:    11035      2107         5535     -1     1     0      0
2019-05-08 18:51:08,460 - INFO #   18 -   19:    11205    -20342        12707     -3    -1    -1      1
2019-05-08 18:51:08,467 - INFO #   20 -   23:    11192    -38017        10051     -5     0    -3      0
2019-05-08 18:51:08,474 - INFO #   24 -   31:    11314    -16475         9987     -4     0    -1      0
2019-05-08 18:51:08,480 - INFO #   32 -   47:    11047    -14494         3447     -2    -1    -1      0
2019-05-08 18:51:08,487 - INFO #   48 -   79:    10997    -14496         3497     -2    -1    -1      0
2019-05-08 18:51:08,494 - INFO #   80 -  143:     8622    -11323         2701     -2    -1    -1      0
2019-05-08 18:51:08,501 - INFO #       TOTAL     75412   -113040       126066     -5     1    -1      1
2019-05-08 18:51:08,504 - INFO # === timer_set_absolute resched ===
2019-05-08 18:51:08,510 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,517 - INFO #   16 -   17:    11396      2085         5686     -1     1     0      0
2019-05-08 18:51:08,524 - INFO #   18 -   19:    11264    -20355        12599     -3    -1    -1      1
2019-05-08 18:51:08,531 - INFO #   20 -   23:    11183    -37955         9996     -5     0    -3      0
2019-05-08 18:51:08,537 - INFO #   24 -   31:    11150    -15990         9234     -4     0    -1      0
2019-05-08 18:51:08,544 - INFO #   32 -   47:    11148    -14626         3477     -2    -1    -1      0
2019-05-08 18:51:08,551 - INFO #   48 -   79:    11203    -14740         3537     -2    -1    -1      0
2019-05-08 18:51:08,557 - INFO #   80 -  143:     8554    -11197         2643     -2    -1    -1      0
2019-05-08 18:51:08,564 - INFO #       TOTAL     75898   -112778       125167     -5     1    -1      1
2019-05-08 18:51:08,567 - INFO # === timer_set_absolute stopped ===
2019-05-08 18:51:08,574 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,581 - INFO #   16 -   17:    11075    -71234         7976     -7    -3    -6      0
2019-05-08 18:51:08,587 - INFO #   18 -   19:    11219    -64546        12172     -6    -3    -5      1
2019-05-08 18:51:08,594 - INFO #   20 -   23:    11149    -42669         9900     -5    -3    -3      0
2019-05-08 18:51:08,601 - INFO #   24 -   31:    11213    -44852            0     -4    -4    -4      0
2019-05-08 18:51:08,607 - INFO #   32 -   47:    11125    -44500            0     -4    -4    -4      0
2019-05-08 18:51:08,614 - INFO #   48 -   79:    11049    -44196            0     -4    -4    -4      0
2019-05-08 18:51:08,621 - INFO #   80 -  143:     8503    -34012            0     -4    -4    -4      0
2019-05-08 18:51:08,628 - INFO #       TOTAL     75333   -346009       238791     -7    -3    -4      3
2019-05-08 18:51:08,632 - INFO # === timer_set_absolute stopped, resched ===
2019-05-08 18:51:08,638 - INFO #    interval     count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,645 - INFO #   16 -   17:    11216    -71932         7984     -7    -3    -6      0
2019-05-08 18:51:08,652 - INFO #   18 -   19:    11056    -63563        11958     -6    -3    -5      1
2019-05-08 18:51:08,658 - INFO #   20 -   23:    11125    -42607         9810     -5    -3    -3      0
2019-05-08 18:51:08,665 - INFO #   24 -   31:    11195    -44780            0     -4    -4    -4      0
2019-05-08 18:51:08,672 - INFO #   32 -   47:    11194    -44776            0     -4    -4    -4      0
2019-05-08 18:51:08,679 - INFO #   48 -   79:    11162    -44648            0     -4    -4    -4      0
2019-05-08 18:51:08,685 - INFO #   80 -  143:     8555    -34220            0     -4    -4    -4      0
2019-05-08 18:51:08,692 - INFO #       TOTAL     75503   -346526       238193     -7    -3    -4      3
2019-05-08 18:51:08,696 - INFO # ===== introspective statistics =====
2019-05-08 18:51:08,699 - INFO # Limits: mean: [-10, 10], variance: [0, 16]
2019-05-08 18:51:08,708 - INFO # self-referencing error (TUT time elapsed - expected TUT interval), in timer under test ticks
2019-05-08 18:51:08,717 - INFO # positive: timer target handling is slow, negative: TUT is dropping ticks or triggering callback early
2019-05-08 18:51:08,726 - INFO # function              count       sum       sum_sq    min   max  mean  variance
2019-05-08 18:51:08,726 - INFO # 
2019-05-08 18:51:08,731 - INFO #          timer_set   301299  -1600719      1726268     -9     0    -5      5
2019-05-08 18:51:08,739 - INFO #            running    75621   -237355       124144     -6     0    -3      1
2019-05-08 18:51:08,746 - INFO #            resched    75427   -237137       123141     -6     0    -3      1
2019-05-08 18:51:08,753 - INFO #            stopped    75090   -562967        87965     -9    -5    -7      1
2019-05-08 18:51:08,760 - INFO #   stopped, resched    75161   -563260        87548     -9    -5    -7      1
2019-05-08 18:51:08,761 - INFO # 
2019-05-08 18:51:08,768 - INFO # timer_set_absolute   302146  -2179831        87681     -9    -4    -7      0
2019-05-08 18:51:08,775 - INFO #            running    75412   -523340       170527     -9    -4    -6      2
2019-05-08 18:51:08,782 - INFO #            resched    75898   -525846       169204     -9    -4    -6      2
2019-05-08 18:51:08,790 - INFO #            stopped    75333   -564753        87902     -9    -5    -7      1
2019-05-08 18:51:08,797 - INFO #   stopped, resched    75503   -565892        87681     -9    -5    -7      1
2019-05-08 18:51:08,801 - INFO # -------------- END STATISTICS ---------------
```

fedepell commented:
@dylad : thanks for checking! The main goal of this PR was to make the test actually work as with the master this will just hang forever on SAML21 and other SAM boards. This can be useful at very least as a regression analysis when touching timer code (ie. we spotted a problem in #11336 with this)
Of course analyzing the numbers in detail is another set of problems :)


I see that for SAML10:
`
2019-05-08 18:45:56,508 - INFO # Context before hardfault:
`
So something is not going well :( It still not block totally (would otherwise even before) but doesn't definitely work, and this is not good.

For SAML21 you confirm it works, so good for that! (if you try master it should not, just hand indefinitely)

I don't know if you have time to dig into why SAML10 blocks (I don't have a hardware to try on so very hard :( ). Otherwise I could remove that from the PR and we see if we make it get merged for the rest?

dylad commented:
@fedepell I'll try to hunt the bug for SAML1X families but maybe we should wait for #11336 before ?

> Otherwise I could remove that from the PR and we see if we make it get merged for the rest?

Since you modified the test app and not something related to SAM0 only, this PR must be tested against several MCU and arch not only SAM0. We might need some help here.

dylad commented:
@fedepell now that #11336 you can slightly rework this PR I think. timer_init() will now ensure timer_stop() is called.
Feels free to rebase it, I'll give it a try on my boards

fedepell commented:
@dylad : rebased to current master, so tested against latest timer unification from @benpicco . Removed now useless timer_stop() indeed, so the PR becomes less invasive.
Thanks if you can test on more hardware, I just have the saml21-xpro from this family.

dylad commented on tests/bench_timers/Makefile:
```diff
@@ -8,6 +8,11 @@ SINGLE_TIMER_BOARDS = \
   native \
   nucleo-f031k6 \
   nucleo-f042k6 \
+  samd21-xpro \
```
samd21-xpro has two timers declared, unless we have an issue here you can remove it.

fedepell replied:
Done!

dylad commented:
SAML10 and SAML11 still hardfault. I'll investigate further.

dylad commented:
Bug has been identify and is not related to this PR. I'll provide a PR to fix it quickly so we can finally move forward here.
